File: db-query-backend/src/llm/openrouter.provider.ts
// import fetch from 'node-fetch';
import {
  LLMProvider,
  GenerateSQLInput,
  GenerateSQLOutput,
} from './llm.types';
import { buildNLtoSQLPrompt } from '../nl/prompt.builder';

export class OpenRouterProvider implements LLMProvider {
  private readonly url = 'https://openrouter.ai/api/v1/chat/completions';

  async generateSQL(
    input: GenerateSQLInput,
  ): Promise<GenerateSQLOutput> {
    const apiKey = process.env.OPENROUTER_API_KEY;
    const model =
      process.env.OPENROUTER_MODEL ??
      'meta-llama/llama-3.1-8b-instruct';

    if (!apiKey) {
      throw new Error('OPENROUTER_API_KEY not set');
    }

    const prompt = buildNLtoSQLPrompt(
      input.dbType,
      input.schema,
      input.question,
    );

    const res = await fetch(this.url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
        'HTTP-Referer': 'http://localhost',
        'X-Title': 'DB-Chatbot',
      },
      body: JSON.stringify({
        model,
        temperature: 0,
        messages: [
          {
            role: 'system',
            content:
              'You generate SQL queries only. No explanations.',
          },
          { role: 'user', content: prompt },
        ],
      }),
    });

    if (!res.ok) {
      throw new Error(`OpenRouter error: ${await res.text()}`);
    }

    const data: any = await res.json();
    const text =
      data?.choices?.[0]?.message?.content?.trim();

    if (!text) {
      throw new Error('OpenRouter returned empty response');
    }

    return {
      sql: text,
      raw: JSON.stringify(data),
      provider: 'openrouter',
    };
  }
}

File: db-query-backend/app/llm/openrouter_provider.py
"""Port of llm/openrouter.provider.ts"""
import httpx

from app.config import get_settings
from app.llm.base import LLMProvider, GenerateSQLInput, GenerateSQLOutput
from app.llm.prompt_builder import build_nl_to_sql_prompt

settings = get_settings()


class OpenRouterProvider(LLMProvider):
    async def generate_sql(self, input: GenerateSQLInput) -> GenerateSQLOutput:
        if not settings.openrouter_api_key:
            raise RuntimeError("OPENROUTER_API_KEY is not set")

        prompt = build_nl_to_sql_prompt(input.db_type, input.schema, input.question)

        async with httpx.AsyncClient(timeout=settings.llm_timeout_ms / 1000) as client:
            resp = await client.post(
                "https://openrouter.ai/api/v1/chat/completions",
                headers={"Authorization": f"Bearer {settings.openrouter_api_key}"},
                json={
                    "model": settings.openrouter_model,
                    "messages": [
                        {
                            "role": "system",
                            "content": "You generate SQL queries only. No explanations.",
                        },
                        {"role": "user", "content": prompt},
                    ],
                },
            )
            resp.raise_for_status()
            data = resp.json()

        raw = data["choices"][0]["message"]["content"]
        return GenerateSQLOutput(sql=raw.strip(), raw=raw, provider="openrouter")
